### PR TITLE
React to #158 (DataToken for GetVirtualPath) in Routing

### PR DESCRIPTION
**Pull request by ianhong**
@rynowak, @Eilon 

This is the PR for the react to https://github.com/aspnet/Routing/pull/158. 

Thanks

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/UrlHelper.cs [hunk 1]
```diff
@@ -99,12 +99,16 @@ private string GeneratePathFromRoute(IDictionary<string, object> values)
...
-            var path = _router.GetVirtualPath(context);
-            if (path == null)
+            var pathData = _router.GetVirtualPath(context);
+            if (pathData == null)
             {
                 return null;
             }
 
+            // VirtualPathData.VirtualPath returns string.Empty for null.
+            Debug.Assert(pathData.VirtualPath != null);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/UrlHelper.cs]: This is going to become invalid with the next change (`PathString` is a struct), your call if you want to take it out now or then.

File: test/Microsoft.AspNet.Mvc.Core.Test/Routing/InnerAttributeRouteTest.cs [hunk 2]
```diff
@@ -417,7 +421,8 @@ public void AttributeRoute_GenerateLink_RespectsPrecedence(string firstTemplate,
         [InlineData("template/{parameter:int:range(1,20)?}", "template", null)]
         [InlineData("template/{parameter:int:range(1,20)?}", "template/5", 5)]
         [InlineData("template/{parameter:int:range(1,20)?}", null, 21)]
-        public void AttributeRoute_GenerateLink_OptionalInlineParameter(string template, string expectedResult, object parameter)
+        public void AttributeRoute_GenerateLink_OptionalInlineParameter
+            (string template, string expectedPath, object parameter)
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/Routing/InnerAttributeRouteTest.cs]: linebreak after the paren

**Comment by rynowak** [comment]
:shipit:

**Comment by Eilon** [comment]
Looks good to me too.

**Comment by ianhong** [comment]
https://github.com/aspnet/Mvc/commit/27bdec40a6a26c83e8219b94f99c692f15dd126f

and 
3693002b0234a7df2c08be694b1514426e057c16

**Pull request closed without merging**